compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/RemotePackageRepository.java

```diff
@@ -284,27 +287,116 @@ public Collection<PackageMetadataResponse> getPackageMetadata(Collection<Resolut
                 }
             }
         }
+
+        List<PackageMetadataResponse> remotePackages = new ArrayList<>(retrieveFromCache(updatedRequests));
+
         // Resolve the requests from remote repository if there are unresolved requests
         if (!updatedRequests.isEmpty()) {
             try {
                 PackageResolutionRequest packageResolutionRequest = toPackageResolutionRequest(updatedRequests);
-                Collection<PackageMetadataResponse> remotePackages =
-                        fromPackageResolutionResponse(updatedRequests, packageResolutionRequest);
-                // Merge central requests and local requests
-                // Here we will pick the latest package from remote or local
-                return mergeResolution(remotePackages, cachedPackages, deprecatedPackages);
-
+                remotePackages.addAll(fromPackageResolutionResponse(updatedRequests, packageResolutionRequest));
             } catch (ConnectionErrorException e) {
                 // ignore connect to remote repo failure
                 // TODO we need to add diagnostics for resolution errors
             } catch (CentralClientException e) {
                 throw new ProjectException(e.getMessage());
             }
         }
+        if (!remotePackages.isEmpty()) {
+            // Merge central requests and local requests
+            // Here we will pick the latest package from remote or local
+            return mergeResolution(remotePackages, cachedPackages, deprecatedPackages);
+        }
         // Return cachedPackages when central requests are not performed
         return cachedPackages;
     }
 
+    private Collection<PackageMetadataResponse> retrieveFromCache(List<ResolutionRequest> updatedRequests) {
+        List<PackageMetadataResponse> response = new ArrayList<>();
+        Iterator<ResolutionRequest> requestIterator = updatedRequests.iterator();
+        while (requestIterator.hasNext()) {
+            ResolutionRequest request = requestIterator.next();
+            for (Map.Entry<ResolutionRequest, PackageMetadataResponse> entry : cachedPackageMetadata.entrySet()) {
+
+                ResolutionRequest cachedRequest = entry.getKey();
+                PackageMetadataResponse pkgMetaDataResponse = entry.getValue();
+                PackageDescriptor latestPkgDesc = pkgMetaDataResponse.resolvedDescriptor();
+
+                if (!(cachedRequest.orgName().equals(request.orgName()) &&
+                        cachedRequest.packageName().equals(request.packageName()))) {
+                    continue;
+                }
+                if (cachedRequest.version().isPresent() && request.version().isPresent()) {
+                    // Check in both requests
+                    if (cachedRequest.version().get().value().major() == request.version().get().value().major() &&
```

Comment:
Can we get values like `cachedRequest.version().get().value().major()` into variables and evaluate the if conditions so that the code looks more readable?